modules/cmr_module/GranuleUMM.cc

```diff
@@ -0,0 +1,242 @@
+
+// Copyright (c) 2015 OPeNDAP, Inc.
+// Author: Nathan Potter <ndp@opendap.org>
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+//
+// You can contact OPeNDAP, Inc. at PO Box 112, Saunderstown, RI. 02874-0112.
+
+/*
+ * Granule.h
+ *
+ *  Created on: July, 13 2018
+ *      Author: ndp
+ */
+#include "config.h"
+
+#include <sstream>
+#include <algorithm>
+#include <string>
+
+#include "nlohmann/json.hpp"
+
+#include "BESDebug.h"
+#include "BESUtil.h"
+
+#include "CmrNames.h"
+#include "CmrApi.h"
+#include "CmrInternalError.h"
+#include "GranuleUMM.h"
+
+
+using namespace std;
+
+#define prolog std::string("GranuleUMM::").append(__func__).append("() - ")
+
+
+namespace cmr {
+
+
+GranuleUMM::GranuleUMM(const nlohmann::json& granule_json)
+{
+    BESDEBUG(MODULE, prolog << "BEGIN" << endl << granule_json.dump(2) << endl);
+    setConceptId(granule_json);
+    setName(granule_json);
+    setSize(granule_json);
+    setDapServiceUrl(granule_json);
+    setDataGranuleUrl(granule_json);
+    setLastModifiedStr(granule_json);
+    setDescription(granule_json);
+}
+
+
+void GranuleUMM::setName(const nlohmann::json& j_obj) {
+    CmrApi cmrApi;
+    const auto &umm_obj = cmrApi.qc_get_object(CMR_UMM_UMM_KEY, j_obj);
+    this->d_name = cmrApi.get_str_if_present(CMR_UMM_GRANULE_UR_KEY, umm_obj);
+}
+
+void GranuleUMM::setDescription(const nlohmann::json& go){
+}
+
+void GranuleUMM::setConceptId(const nlohmann::json& j_obj)
+{
+    CmrApi cmrApi;
+    const auto &meta_obj = cmrApi.qc_get_object(CMR_UMM_META_KEY, j_obj);
+
+    BESDEBUG(MODULE, prolog << "META OBJECT" << endl << cmrApi.probe_json(meta_obj) << endl);
+
+    this->d_id = cmrApi.get_str_if_present(CMR_UMM_CONCEPT_ID_KEY, meta_obj);
+}
+
+
+void GranuleUMM::setSize(const nlohmann::json& granule_obj)
+{
+    CmrApi cmrApi;
+    const auto &umm_obj = cmrApi.qc_get_object(CMR_UMM_UMM_KEY, granule_obj);
+    const auto &data_granule_obj = cmrApi.qc_get_object(CMR_UMM_DATA_GRANULE_KEY, umm_obj);
+    BESDEBUG(MODULE, prolog << CMR_UMM_DATA_GRANULE_KEY << data_granule_obj.dump(2) << endl );
+    const auto &arch_and_info_array = cmrApi.qc_get_array(CMR_UMM_ARCHIVE_AND_DIST_INFO_KEY, data_granule_obj);
+    for(const auto &entry : arch_and_info_array){
```

Comment:
This should be re-written as an if since it will only be run once.

Reply:
It _might_ be run once. 
There is a need to process an array of objects and an "if" is not really optimal for that

Reply:
See the comment I added to the code. Although in this incarnation it does get runonce I have counter examples where it may have to run more than once. At this point I would say the determination of size is a bit of a punt - the CMR records are both unreliably populated and difficult to discriminate. 🤷‍♂️